Update a batch request's txn from its response's in DistSender.sendChunk

kkaneda opened the pull request:
The issue was initially discussed in #3112. 

This commits adds two test cases. The first one tests a fix made by this commit. The second one reproduces a case where txn data are not still propagated correctly.

The second test case is interesting. A different txn ID is assigned whenever a txn is restarted, so a write made by the same txn of two different epochs can cause a write conflict error.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3161)

<!-- Reviewable:end -->

kkaneda committed:
Add RangeKeyMismatchError to SequenceCache.shouldCacheError

tschottdorf commented:
Reviewed 1 of 2 files at r1.
Review status: 1 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

_[kv/dist_sender_server_test.go, line 423 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K3VUaiBzK33Z1nqY-D6:-K3VUaiBzK33Z1nqY-D7:-1816426609) ([raw file](https://github.com/cockroachdb/cockroach/blob/b02e7b530371c90c2b5e4079917269af828693c9/kv/dist_sender_server_test.go#L423)):_
mention that you're really testing propagation of the Txn in general, but that the Writing flag is convenient to test for here.

---

_[kv/dist_sender_server_test.go, line 440 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K3VTQ9LvccBJ4cTEEhc:-K3VTQ9LvccBJ4cTEEhd:2129782065) ([raw file](https://github.com/cockroachdb/cockroach/blob/b02e7b530371c90c2b5e4079917269af828693c9/kv/dist_sender_server_test.go#L440)):_
is that really what's happening? You haven't written to that replica at all (your previous attempt got the mismatch error, which should happen before executing the cmd), so the sequence cache there ought to be empty unless you're talking to the first replica again.

---

_[kv/dist_sender_server_test.go, line 462 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K3VUQ1zNNtFqr50B7zG:-K3VUQ1zNNtFqr50B7zH:278058231) ([raw file](https://github.com/cockroachdb/cockroach/blob/b02e7b530371c90c2b5e4079917269af828693c9/kv/dist_sender_server_test.go#L462)):_
if you expect an error, assert on it here. If we stick with the TransactionRetryError, you should verify that the contained transaction has the updated from the first range (i.e. the Writing flag).

---

_[kv/dist_sender_server_test.go, line 475 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K3VVPSIU47GKVgsy8fb:-K3VVPSIU47GKVgsy8fc:302774834) ([raw file](https://github.com/cockroachdb/cockroach/blob/b02e7b530371c90c2b5e4079917269af828693c9/kv/dist_sender_server_test.go#L475)):_
add an issue number and mention that the underlying issue is that not all errors carry a transaction, but range-spanning writes are not atomic so data may actually have been written.

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161)_
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 2 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161)_
<!-- Sent from Reviewable.io -->

kkaneda commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

_[kv/dist_sender_server_test.go, line 423 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K3VUaiBzK33Z1nqY-D6:-K3_LNS1QvPDaVAnd0PI:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b02e7b530371c90c2b5e4079917269af828693c9/kv/dist_sender_server_test.go#L423)):_
Done.

---

_[kv/dist_sender_server_test.go, line 440 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K3VTQ9LvccBJ4cTEEhc:-K3_LON5yAEuJkDsYS-5:671999756) ([raw file](https://github.com/cockroachdb/cockroach/blob/b02e7b530371c90c2b5e4079917269af828693c9/kv/dist_sender_server_test.go#L440)):_
Yes, I think so. We have two places to check range mismatch (`LocalSender.lookupReplica` and `Replica.checkCmdHeader`), and we hit the latter in this case. `LocalSender.Send` didn't call `lookupReplica` since `RangeID` and `Replica.StoreID` were specified.

`Replica.checkCmdHeader` returned an error when executing the put command on "c", which was after the put command on "b" was executed.

---

_[kv/dist_sender_server_test.go, line 462 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K3VUQ1zNNtFqr50B7zG:-K3_M5abWq3W8CCSKJPI:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b02e7b530371c90c2b5e4079917269af828693c9/kv/dist_sender_server_test.go#L462)):_
Done.

---

_[kv/dist_sender_server_test.go, line 475 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K3VVPSIU47GKVgsy8fb:-K3_M6MXK46YsSozNeQk:258837848) ([raw file](https://github.com/cockroachdb/cockroach/blob/b02e7b530371c90c2b5e4079917269af828693c9/kv/dist_sender_server_test.go#L475)):_
Sure, added a reference to #743.

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161)_
<!-- Sent from Reviewable.io -->

kkaneda commented:
Thanks for the review, @tschottdorf ! Addressed your comments.

tschottdorf commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

_[kv/dist_sender_server_test.go, line 440 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K3VTQ9LvccBJ4cTEEhc:-K3_UBsWcFb-iWZDlOKJ:-737951405) ([raw file](https://github.com/cockroachdb/cockroach/blob/b02e7b530371c90c2b5e4079917269af828693c9/kv/dist_sender_server_test.go#L440)):_
I see. It's irrelevant that `Put(b)` executes earlier, though - what matters is that we populate the sequence cache on this particular error. Could you add `RangeKeyMismatchError` to the error types excluded in `shouldCacheError` of the sequence cache? That should alter the behavior of the test.

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161)_
<!-- Sent from Reviewable.io -->

kkaneda commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

_[kv/dist_sender_server_test.go, line 440 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K3VTQ9LvccBJ4cTEEhc:-K3_ZG9Kc0LN-R0aTDat:-1383137102) ([raw file](https://github.com/cockroachdb/cockroach/blob/b02e7b530371c90c2b5e4079917269af828693c9/kv/dist_sender_server_test.go#L440)):_
Ah, I see. Yes, the change the behavior of the test (no longer retry). Let me rework on the test. I can send a separate PR for the change to `shouldCacheError` if it's better (since rewriting the test might take some time..)

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161)_
<!-- Sent from Reviewable.io -->

tschottdorf commented:
It's fine with me if you just add a commit for `shouldCacheError` and rebase that under the current code in the PR. It's not really blocking anything.

kkaneda commented:
Sure, will update the PR when the test is updated. Thanks!

kkaneda committed:
Update a batch request's txn from its response's in DistSender.sendChunk

The issue was initially discussed in #3112. This commits adds two test
cases. The first one tests a fix made by this commit. The second one
reproduces a case where txn data are not still propagated correctly.

The second test case is interesting. A different txn ID is assigned
whenever a txn is restarted, so a write made by the same txn of two
different epochs can cause a write conflict error.

kkaneda commented:
Updated the test. @tschottdorf , could you take another look?

tschottdorf commented:
LGTM. Thanks Kenji!

---

Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

_[kv/dist_sender.go, line 660 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K3oqOAVEXfNZUzt7G06:-K3oqOAVEXfNZUzt7G07:-863954412) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfe44e73c6ab99cd3f169c2a92a5b32167cd2d49/kv/dist_sender.go#L660)):_
is this really needed? Updating (a possibly `nil` *Transaction`) from a `nil` `*Transaction` should work fine (?).

---

_[kv/dist_sender_server_test.go, line 426 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K3orth1lI5-a6AQ5Hkd:-K3orth1lI5-a6AQ5Hke:1501333918) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfe44e73c6ab99cd3f169c2a92a5b32167cd2d49/kv/dist_sender_server_test.go#L426)):_
consider renaming to `TestNoSequenceCachePutOnRangeMismatchError` and add that this is a higher-level version of `TestSequenceCacheShouldCache`.

---

_[kv/dist_sender_server_test.go, line 472 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K3or2a_LGyfXivfUkUv:-K3or2aakPsx9QRSUsxx:295890555) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfe44e73c6ab99cd3f169c2a92a5b32167cd2d49/kv/dist_sender_server_test.go#L472)):_
you need to set and deferred-undo `TestingCommandFilter` before you have any moving parts, i.e. before the call to `setupMultipleRanges`.

---

_[kv/dist_sender_server_test.go, line 494 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K3orDe33ur5tgQIbLSW:-K3orDe33ur5tgQIbLSX:1278673161) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfe44e73c6ab99cd3f169c2a92a5b32167cd2d49/kv/dist_sender_server_test.go#L494)):_
s/are/is/

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-:-K3osFeunQ-gOcS0T8L0:-1674343462)_
<!-- Sent from Reviewable.io -->

kkaneda commented:
Thanks for the quick review!

---

Review status: 2 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

_[kv/dist_sender.go, line 660 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K3oqOAVEXfNZUzt7G06:-K3ozyLSR_pQ9DfcPCbL:1271246679) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfe44e73c6ab99cd3f169c2a92a5b32167cd2d49/kv/dist_sender.go#L660)):_
Ah, you're right. Not needed. Fixed.

---

_[kv/dist_sender_server_test.go, line 426 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K3orth1lI5-a6AQ5Hkd:-K3p-3PJoLObgGWZFzvC:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfe44e73c6ab99cd3f169c2a92a5b32167cd2d49/kv/dist_sender_server_test.go#L426)):_
Done.

---

_[kv/dist_sender_server_test.go, line 472 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K3or2a_LGyfXivfUkUv:-K3p-23xL7ntlhDXs17Y:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfe44e73c6ab99cd3f169c2a92a5b32167cd2d49/kv/dist_sender_server_test.go#L472)):_
Done.

---

_[kv/dist_sender_server_test.go, line 494 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K3orDe33ur5tgQIbLSW:-K3p-2d_g31wfhdbvfG4:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfe44e73c6ab99cd3f169c2a92a5b32167cd2d49/kv/dist_sender_server_test.go#L494)):_
Done.

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-:-K3p-43SkRk90s1Lr50U:-1411669730)_
<!-- Sent from Reviewable.io -->

tschottdorf commented:
LGTM. Interestingly, your test failed, though :-)

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-:-K3p1jD4qXibRFCpsshb:187614865)_
<!-- Sent from Reviewable.io -->

kkaneda commented:
Yes... It seems the sequence cache was updated for some reason... Trying to reproduce..

kkaneda commented:
Reproduced a test failure by calling `Time.sleep` from an async task created in `Replica.handleSkippedIntents`. When an intent is not resolved after split, the range descriptor cache is not able to get the correct descriptor of the replica storing key "c". Thus, even after the initial `RangeKeyMismatchError` is returned, we end up ending two batch requests (put on "b" and put on "c") to the same replica storing key "b". The first request updates the sequence cache, and then the second request will fail.

Working on a fix...

tschottdorf commented:
The fix is moving the sequence counter increment from chunkingSender to
reach range query. See #3206.
On Nov 25, 2015 3:01 AM, "Kenji Kaneda" <notifications@github.com> wrote:

> Reproduced a test failure by calling Time.sleep from an async task
> created in Replica.handleSkippedIntents. When an intent is not resolved
> after split, the range descriptor cache is not able to get the correct
> descriptor of the replica storing key "c". Thus, even after the initial
> RangeKeyMismatchError is returned, we end up ending two batch requests
> (put on "b" and put on "c") to the same replica storing key "b". The first
> request updates the sequence cache, and then the second request will fail.
>
> Working on a fix...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/3161#issuecomment-159528817>
> .
>

kkaneda committed:
Move SetNewRequest() from chunkingSender.Send to DistSender.sendAttempt

This change addresses an infinite loop issue discussed in #3206. More
specifically, DistSender might send multiple batch requests to the
same replica when it finds a post-split range descriptor in the cache
while the split has not yet been fully completed. By giving a higher
sequence to the second request, we can avoid an infinite txn restart
error by hitting the sequence cache.

kkaneda commented:
Ah, I see. Thanks, Tobias. The test passed if we move `ba.SetNewRequest()` from `chunkingSender.Send` to `Stores.Send`. Let me try writing a test (without adding a sleep to handleSkippedIntents).

tschottdorf commented:
You're welcome, thanks for dealing with this. The test is appreciated.

On Wed, Nov 25, 2015 at 9:44 AM, Kenji Kaneda <notifications@github.com>
wrote:

> Ah, I see. Thanks, Tobias. The test passed if we move ba.SetNewRequest()
> from chunkingSender.Send to Stores.Send. Let me try writing a test
> (without adding a sleep to handleSkippedIntents).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/3161#issuecomment-159627873>
> .
>

kkaneda commented:
Sure! Added a test. It's a bit awkward, but reproduces the bug by making ResolveIntentRequest always fail. Could you take a look?

tschottdorf commented:
I'm a little confused, commented on the tests. Let me know if I'm on the wrong track.

---

Reviewed 6 of 6 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[kv/dist_sender_server_test.go, line 627 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K3zwR9wg9AOaJ5l7L7l:-K3zwR9wg9AOaJ5l7L7m:1945031846) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec286cb9a21f06cb6c4e9bd9276bb6d34e9ce92/kv/dist_sender_server_test.go#L627)):*
could you rewrite that last sentence, removing `chunked`? `chunking`" is what `chunkingSender` does (not that we define that anywhere, but still). Instead, explain that the issue is that a split intent descriptor can make it into the descriptor cache, leading to querying a range as two ranges when it really hasn't split yet.

---

*[kv/dist_sender_server_test.go, line 658 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K3zy7irUeTTOfy0wjVJ:-K3zy7irUeTTOfy0wjVK:1788802080) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec286cb9a21f06cb6c4e9bd9276bb6d34e9ce92/kv/dist_sender_server_test.go#L658)):*
I don't understand what's going on here. Looks like with your setup, you're going to see the new descriptors regardless - a `RangeKeyMismatchError` triggers `considerIntents`, so your requests will retry at the `DistSender` until all the intents have entered the cache. Can you really guarantee the stale descriptor? See also below about what we actually want to test.

---

*[kv/dist_sender_server_test.go, line 659 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K3zxfi27X-h6M3H3jlR:-K3zxfi27X-h6M3H3jlS:148199914) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec286cb9a21f06cb6c4e9bd9276bb6d34e9ce92/kv/dist_sender_server_test.go#L659)):*
This doesn't seem like we're testing the right thing. I want a range which hasn't split yet, but the range descriptor cache should already think the split had happened, so that when it addresses the second half of the range separately, there's a restart (and this should be fixed by moving the call to `ba.SetNewRequest`.

I also think we should modify the descriptor cache directly for this test, which suggests having it not in the server test (but instead just testing `DistSender` in isolation, verifying in the `rpcSend` mock that the sequence number has increased between calls to both halves of our single range.

---

*[storage/stores.go, line 153 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K3zwzxzL0fb4iHvVC8f:-K3zwzxzL0fb4iHvVC8g:-1182048346) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec286cb9a21f06cb6c4e9bd9276bb6d34e9ce92/storage/stores.go#L153)):*
you can't do that here, it must be in the `DistSender` multi-range-query loop. If you do it here, you've effectively disabled replay protection.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-:-K3zyuwSK-Dbfsyg0XQu:619449843)*
<!-- Sent from Reviewable.io -->

kkaneda commented:
Thanks for the detailed explanation. Sorry, I was sending the PR in too haste... Reworked.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[kv/dist_sender_server_test.go, line 627 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K3zwR9wg9AOaJ5l7L7l:-K40YMl79hJLG0hyZvm4:1856916990) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec286cb9a21f06cb6c4e9bd9276bb6d34e9ce92/kv/dist_sender_server_test.go#L627)):*
Sure. Updated the description as you suggested (and since the test is now a unittest that checks sequence update).

```
// TestSequenceUpdateOnMultiRangeQueryLoop reproduces #3206 and
// verifies that the sequence is updated in the DistSender
// multi-range-query loop.
//
// More specifically, the issue was that DistSender might send
// multiple batch requests to the same replica when it finds a
// post-split range descriptor in the cache while the split has not
// yet been fully completed. By giving a higher sequence to the second
// request, we can avoid an infinite txn restart error by hitting the
// sequence cache.
```

---

*[kv/dist_sender_server_test.go, line 658 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K3zy7irUeTTOfy0wjVJ:-K40Z-_ytYmj6qZBRixl:650054448) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec286cb9a21f06cb6c4e9bd9276bb6d34e9ce92/kv/dist_sender_server_test.go#L658)):*
You're right. I think I was looking at something different. Fixed.

---

*[kv/dist_sender_server_test.go, line 659 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K3zxfi27X-h6M3H3jlR:-K40Z5Fcz4nnOSELLfA1:-1264057523) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec286cb9a21f06cb6c4e9bd9276bb6d34e9ce92/kv/dist_sender_server_test.go#L659)):*
Got it. Yes, using mock would be better.. Fixed.

---

*[storage/stores.go, line 153 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K3zwzxzL0fb4iHvVC8f:-K40ZA6Zr76lHglN9GS_:1246726980) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec286cb9a21f06cb6c4e9bd9276bb6d34e9ce92/storage/stores.go#L153)):*
Ah, I see. Moved the code to `DistSender.sendAttempt`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-:-K40YBBnyDwZXe1xxwCn:-1161913000)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
LGTM. Last round of nits and off we go. Thanks as always Kenji!

---

Reviewed 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[kv/dist_sender.go, line 437 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K4OatHz2IClZcFy9Qm5:-K4OatHz2IClZcFy9Qm6:1215535474) ([raw file](https://github.com/cockroachdb/cockroach/blob/95798a32f12b8344b35070c45b4601318cd6657b/kv/dist_sender.go#L437)):*
even if it were impossible to hit the same range twice, we'd have to increase the counter here. Probably better to explain that we do it in the per-range loop (and not outside) to account for the fact that we **might** hit the same range twice "by accident" (I wouldn't call this "likely").

---

*[kv/dist_sender_test.go, line 968 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K4ObZBYA8FXPj7HFsg_:-K4ObZBYA8FXPj7HFsga:1382212968) ([raw file](https://github.com/cockroachdb/cockroach/blob/95798a32f12b8344b35070c45b4601318cd6657b/kv/dist_sender_test.go#L968)):*
> we can avoid an infinite txn restart (otherwise caused by hitting the sequence cache).

---

*[kv/dist_sender_test.go, line 1042 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K4Oc0lLo2LSilcNo1gV:-K4Oc0lLo2LSilcNo1gW:395411568) ([raw file](https://github.com/cockroachdb/cockroach/blob/95798a32f12b8344b35070c45b4601318cd6657b/kv/dist_sender_test.go#L1042)):*
s/exe/exp/

---

*[kv/dist_sender_test.go, line 1047 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K4Oc5oB5aTSSbfwdXAj:-K4Oc5oCshI5kTUpVp8Z:1405270303) ([raw file](https://github.com/cockroachdb/cockroach/blob/95798a32f12b8344b35070c45b4601318cd6657b/kv/dist_sender_test.go#L1047)):*
nit: you can just `return []proto.Message{ba.CreateReply()}`, I think.

---

*[kv/dist_sender_test.go, line 1058 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K4OcIKb9MYpocqHXdzU:-K4OcIKb9MYpocqHXdzV:37476050) ([raw file](https://github.com/cockroachdb/cockroach/blob/95798a32f12b8344b35070c45b4601318cd6657b/kv/dist_sender_test.go#L1058)):*
s/contains/containing/

---

*[kv/dist_sender_test.go, line 1059 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K4OcKf_909DUdWavxLO:-K4OcKf_909DUdWavxLP:-11344330) ([raw file](https://github.com/cockroachdb/cockroach/blob/95798a32f12b8344b35070c45b4601318cd6657b/kv/dist_sender_test.go#L1059)):*
nit: var ba roachpb.BatchRequest

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-:-K4Oc_y9msx2UBg35uy6:733793156)*
<!-- Sent from Reviewable.io -->

kkaneda commented:
Thanks, Tobias! Addressed all comments.

---

Review status: 4 of 6 files reviewed at latest revision, 6 unresolved discussions.

---

*[kv/dist_sender.go, line 437 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K4OatHz2IClZcFy9Qm5:-K4QVe9eviGoz3rJJDpY:-841344948) ([raw file](https://github.com/cockroachdb/cockroach/blob/95798a32f12b8344b35070c45b4601318cd6657b/kv/dist_sender.go#L437)):*
Sure! Updated the description. 

```
	// Increase the sequence counter in the per-range loop (not
	// outside) since we might hit the same range twice by
	// accident.
```

---

*[kv/dist_sender_test.go, line 968 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K4ObZBYA8FXPj7HFsg_:-K4QVrIpgXY4xBYb0wrV:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/95798a32f12b8344b35070c45b4601318cd6657b/kv/dist_sender_test.go#L968)):*
Done.

---

*[kv/dist_sender_test.go, line 1042 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K4Oc0lLo2LSilcNo1gV:-K4QVqeO_cYYDczL-1dk:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/95798a32f12b8344b35070c45b4601318cd6657b/kv/dist_sender_test.go#L1042)):*
Done.

---

*[kv/dist_sender_test.go, line 1047 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K4Oc5oB5aTSSbfwdXAj:-K4QVs532pLAsSlcDvsc:398748410) ([raw file](https://github.com/cockroachdb/cockroach/blob/95798a32f12b8344b35070c45b4601318cd6657b/kv/dist_sender_test.go#L1047)):*
Yes, that worked! Done.

---

*[kv/dist_sender_test.go, line 1058 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K4OcIKb9MYpocqHXdzU:-K4QVuQeoQC5DWU-LFs-:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/95798a32f12b8344b35070c45b4601318cd6657b/kv/dist_sender_test.go#L1058)):*
Done.

---

*[kv/dist_sender_test.go, line 1059 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-K4OcKf_909DUdWavxLO:-K4QVup4Kpcspr5tUgyx:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/95798a32f12b8344b35070c45b4601318cd6657b/kv/dist_sender_test.go#L1059)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3161#-:-K4QW1LlaFc-tI-eKp3D:253670489)*
<!-- Sent from Reviewable.io -->
